File: src/dynamic programming/python/count construct/src/count_construct.py
def count_construct_basic(target, word_bank):

    if target == "":
        return 1

    totalCount = 0

    for word in word_bank:
        if len(target) >= len(word) and target[: len(word)] == word:
            remainder = target[len(word) :]
            totalCount += count_construct_basic(remainder, word_bank)

    return totalCount


def count_construct_memo(target, word_bank, memo=dict()):

    if target == "":
        return 1

    if target in memo:
        return memo[target]

    totalCount = 0

    for word in word_bank:
        if len(target) >= len(word) and target[: len(word)] == word:
            remainder = target[len(word) :]
            totalCount += count_construct_memo(remainder, word_bank, memo)
    
    memo[target] = totalCount
    return totalCount


def count_construct_table(target, word_bank):

    table = [0] * (len(target) + 1)
    table[0] = 1

    for i in range(len(target)):
        if table[i]:
            for word in word_bank:
                if target[i : i + len(word)] == word:
                    table[i + len(word)] += table[i]

    return table[-1]

File: src/dynamic programming/cpp/count construct/src/count_construct.cpp
#include "count_construct.h"
#include <functional>
#include <unordered_map>

int countConstructBasic(const std::string &target,
                        const std::vector<std::string> &wordBank) {
  if (target.empty())
    return 1;

  int totalCount = 0;

  for (const auto &word : wordBank) {

    if (target.size() >= word.size() && target.substr(0, word.size()) == word) {
      auto suffix = target.substr(word.size());
      totalCount += countConstructBasic(suffix, wordBank);
    }
  }

  return totalCount;
}

int countConstructMemo(const std::string &target,
                       const std::vector<std::string> &wordBank) {

  std::function<int(const std::string &, const std::vector<std::string> &,
                    std::unordered_map<std::string, int> &)>
      _countConstructMemo;
  _countConstructMemo = [&](const std::string &target,
                            const std::vector<std::string> &wordBank,
                            std::unordered_map<std::string, int> &memo) -> int {
    if (target.empty())
      return 1;

    if (memo.count(target))
      return memo[target];

    int totalCount = 0;

    for (const auto &word : wordBank) {

      if (target.size() >= word.size() &&
          target.substr(0, word.size()) == word) {
        auto suffix = target.substr(word.size());
        totalCount += _countConstructMemo(suffix, wordBank, memo);
      }
    }

    memo[target] = totalCount;
    return totalCount;
  };

  std::unordered_map<std::string, int> memo;

  return _countConstructMemo(target, wordBank, memo);
}

int countConstructTable(const std::string &target,
                        const std::vector<std::string> &wordBank) {

  std::vector<int> table(target.size() + 1);
  table[0] = 1;

  for (int i = 0; i < target.size(); i++) {

    for (const auto &word : wordBank) {

      if (target.substr(i, word.size()) == word) {

        table[i + word.size()] += table[i];
      }
    }
  }

  return table.back();
}
